This project is intended to give you experience working with pandas.  Due Sept 14th 2018, 5pm EST.
1. read in data/iris.csv, which contains data from the iris data set, https://en.wikipedia.org/wiki/Iris_flower_data_set
    1. How many columns are there?  what data types are the columns?  What are the column names and what do they correspond to?
    2. How many observations (rows) are there?
    3. How many species are included in this data set
2. This data set in our repository was the original data set uploaded to the UC Irvine machine learning repository.  It is known that there are errors in this data.  
    * the 35th row should read [4.9,3.1,1.5,0.2,"setosa"]
    * the 38th row should read [4.9,3.6,1.4,0.1,"setosa"]
Correct these entries in your DataFrame object.
3. Add two new columns to your DataFrame:
    * "Petal ratio" : the ratio of the petal length to petal width
    * "Sepal ratio" : the ratio of the sepal length to sepal width"
4. For each species, determine the mean, median, minimum, maximum and standard deviation of the petal and sepal ratio.  You may wish to group data according to species (see section 2.6.7) and use a descriptive summary statistic (see page 15)

In [101]:
import pandas as pd
import numpy as np
iris = pd.read_csv("data/iris.csv")

df = pd.DataFrame(data= iris)

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
Sepal.Length    150 non-null float64
Sepal.Width     150 non-null float64
Petal.Length    150 non-null float64
Petal.Width     150 non-null float64
Species         150 non-null object
dtypes: float64(4), object(1)
memory usage: 5.9+ KB


In [74]:
df.Species.unique()

array(['setosa', 'versicolor', 'virginica'], dtype=object)

In [75]:
df.describe()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

# Question 1
### Part A
The Iris data contains 5 columns.  Four of the columns are Float64 which provide dimensions for various iris samples.  The fifth column is a text column which contains the species name, as an object, for each of the iris samples.  Column names are as follows: Sepal.Length, Sepal.Width, Petal.Length, Petal.Width, Species.  These column names specify dimensions for two parts common to iris species.
### Part B
Included in the data are 150 samples of iris petals.
### Part C
There are 3 unique species included in the data.  They consist of Setosa, Versicolor, and Virginica.
# Question 2
I used the .iloc and .iat functions to find the cell values for the 35th and 38th columns.  Given that python begins the index at 0 the indeces for the altered rows are 34 and 37 respectively.  Using .iloc, I found the values which needed replacement were in the Petal.Width column for row 34 and in the Sepal.Width and Petal.Length columns for row 37.  These values were corrected using the .iat function, and the result is displayed with the df.head(38) function.

In [76]:
df.iloc[34][0:5]

Sepal.Length       4.9
Sepal.Width        3.1
Petal.Length       1.5
Petal.Width        0.1
Species         setosa
Name: 34, dtype: object

In [77]:
df.iat[34,3]=0.2

In [78]:
df.iloc[34][0:5]

Sepal.Length       4.9
Sepal.Width        3.1
Petal.Length       1.5
Petal.Width        0.2
Species         setosa
Name: 34, dtype: object

In [79]:
df.iloc[37][0:5]

Sepal.Length       4.9
Sepal.Width        3.1
Petal.Length       1.5
Petal.Width        0.1
Species         setosa
Name: 37, dtype: object

In [80]:
df.iat[37,1]=3.6

In [81]:
df.iat[37,2]=1.4

In [82]:
df.iloc[37][0:5]

Sepal.Length       4.9
Sepal.Width        3.6
Petal.Length       1.4
Petal.Width        0.1
Species         setosa
Name: 37, dtype: object

In [85]:
df.head(38)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8          4.0           1.2          0.2  setosa
15           5.7          4.4           1.5          0.4  setosa
16           5.4          3.9           1.3          0.4  setosa
17           5.1          3.5           1.4          0.3  setosa
18           5.7          3.8           1.7          0.3  setosa
19           5.1          3.8           1.5          0.3  setosa
20           5.4          3.4           1.7          0.2  setosa
21           5.1          3.7           1.5          0.4  setosa
22           4.6          3.6           1.0          0.2  setosa
23           5.1          3.3           1.7          0.5  setosa
24           4.8          3.4           1.9          0.2  setosa
25           5.0          3.0           1.6          0.2  setosa
26           5.0          3.4           1.6          0.4  setosa
27           5.2          3.5           1.5          0.2  setosa
28           5.2          3.4           1.4          0.2  setosa
29           4.7          3.2           1.6          0.2  setosa
30           4.8          3.1           1.6          0.2  setosa
31           5.4          3.4           1.5          0.4  setosa
32           5.2          4.1           1.5          0.1  setosa
33           5.5          4.2           1.4          0.2  setosa
34           4.9          3.1           1.5          0.2  setosa
35           5.0          3.2           1.2          0.2  setosa
36           5.5          3.5           1.3          0.2  setosa
37           4.9          3.6           1.4          0.1  setosa

# Question 3

In [92]:
#Petal ratio: Petal Length/Petal Width
df['Petal ratio'] = df['Petal.Length'] / df['Petal.Width']
#Sepal ratio: Sepal Length/Sepal Width
df['Sepal ratio'] = df['Sepal.Length'] / df['Sepal.Width']
df.head(38)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species  Petal ratio  \
0            5.1          3.5           1.4          0.2  setosa     7.000000   
1            4.9          3.0           1.4          0.2  setosa     7.000000   
2            4.7          3.2           1.3          0.2  setosa     6.500000   
3            4.6          3.1           1.5          0.2  setosa     7.500000   
4            5.0          3.6           1.4          0.2  setosa     7.000000   
5            5.4          3.9           1.7          0.4  setosa     4.250000   
6            4.6          3.4           1.4          0.3  setosa     4.666667   
7            5.0          3.4           1.5          0.2  setosa     7.500000   
8            4.4          2.9           1.4          0.2  setosa     7.000000   
9            4.9          3.1           1.5          0.1  setosa    15.000000   
10           5.4          3.7           1.5          0.2  setosa     7.500000   
11           4.8          3.4           1.6          0.2  setosa     8.000000   
12           4.8          3.0           1.4          0.1  setosa    14.000000   
13           4.3          3.0           1.1          0.1  setosa    11.000000   
14           5.8          4.0           1.2          0.2  setosa     6.000000   
15           5.7          4.4           1.5          0.4  setosa     3.750000   
16           5.4          3.9           1.3          0.4  setosa     3.250000   
17           5.1          3.5           1.4          0.3  setosa     4.666667   
18           5.7          3.8           1.7          0.3  setosa     5.666667   
19           5.1          3.8           1.5          0.3  setosa     5.000000   
20           5.4          3.4           1.7          0.2  setosa     8.500000   
21           5.1          3.7           1.5          0.4  setosa     3.750000   
22           4.6          3.6           1.0          0.2  setosa     5.000000   
23           5.1          3.3           1.7          0.5  setosa     3.400000   
24           4.8          3.4           1.9          0.2  setosa     9.500000   
25           5.0          3.0           1.6          0.2  setosa     8.000000   
26           5.0          3.4           1.6          0.4  setosa     4.000000   
27           5.2          3.5           1.5          0.2  setosa     7.500000   
28           5.2          3.4           1.4          0.2  setosa     7.000000   
29           4.7          3.2           1.6          0.2  setosa     8.000000   
30           4.8          3.1           1.6          0.2  setosa     8.000000   
31           5.4          3.4           1.5          0.4  setosa     3.750000   
32           5.2          4.1           1.5          0.1  setosa    15.000000   
33           5.5          4.2           1.4          0.2  setosa     7.000000   
34           4.9          3.1           1.5          0.2  setosa     7.500000   
35           5.0          3.2           1.2          0.2  setosa     6.000000   
36           5.5          3.5           1.3          0.2  setosa     6.500000   
37           4.9          3.6           1.4          0.1  setosa    14.000000   

    Sepal ratio  
0      1.457143  
1      1.633333  
2      1.468750  
3      1.483871  
4      1.388889  
5      1.384615  
6      1.352941  
7      1.470588  
8      1.517241  
9      1.580645  
10     1.459459  
11     1.411765  
12     1.600000  
13     1.433333  
14     1.450000  
15     1.295455  
16     1.384615  
17     1.457143  
18     1.500000  
19     1.342105  
20     1.588235  
21     1.378378  
22     1.277778  
23     1.545455  
24     1.411765  
25     1.666667  
26     1.470588  
27     1.485714  
28     1.529412  
29     1.468750  
30     1.548387  
31     1.588235  
32     1.268293  
33     1.309524  
34     1.580645  
35     1.562500  
36     1.571429  
37     1.361111

# Question 4
I accomplised this by using the .groupby and .describe functions to sort the species and run descriptive statistics on the groups.

In [100]:
setgroup = df[["Species","Petal ratio","Sepal ratio"]].groupby('Species')
descriptgroup = setgroup.describe()
descriptgroup

Petal ratio                                                    \
                 count      mean       std       min       25%       50%   
Species                                                                    
setosa            50.0  6.908000  2.854545  2.666667  4.687500  7.000000   
versicolor        50.0  3.242837  0.312456  2.666667  3.016667  3.240385   
virginica         50.0  2.780662  0.407367  2.125000  2.511364  2.666667   

                           Sepal ratio                                \
                 75%   max       count      mean       std       min   
Species                                                                
setosa      7.500000  15.0        50.0  1.470188  0.118750  1.268293   
versicolor  3.417582   4.1        50.0  2.160402  0.228658  1.764706   
virginica   3.055556   4.0        50.0  2.230453  0.246992  1.823529   

                                                    
                 25%       50%       75%       max  
Species                                             
setosa      1.385684  1.463063  1.541444  1.956522  
versicolor  2.033929  2.161290  2.232692  2.818182  
virginica   2.031771  2.169540  2.342949  2.961538